### ISSUE TITLE
don't emit reportUnknownArgumentType when it doesn't matter

### ISSUE THREAD

**Issue by jab** [microsoft/pyright#4314]
To get a reference certain types like e.g. `dict_keys`, which are not available via builtins or direct import, it is necessary to write code like the following:
```python
dict_keys = type({}.keys())
```
Pyright flags this code with the error `Argument type is partially unknown... Argument type is "dict_keys[Unknown, Unknown]"`.

You can appease Pyright by adding a useless item to the dict, e.g.:
```python
dict_keys = type({0: 0}.keys())
```

But I think it shouldn't be necessary to obscure and bloat the code like this. Ideally, Pyright would detect that the dict's key type and value type are irrelevant in this code, and not complain about it.

For comparison, `mypy --strict` does not emit any errors on this code.

**Comment by erictraut** [comment]
Pyright's behavior here is consistent with its behavior in other cases. Mypy doesn't have the notion of an `Unknown` (an implied `Any`), so that explains the difference here.

Here's how I recommend providing the type arguments in this case:

```python
dict_keys = type(dict[str, str]().keys())
```

**Comment by jab** [comment]
Not sure the request here was clear, so just in case, the proposal here is that Pyright should still flag code where Unknown matters, because knowing the type is actually required to type check subsequent uses. Only when it doesn't matter should Pyright not flag.

So in the example here, merely getting a reference to the `dict_keys` type should not be an error. Only if later code tried to use that reference in a way that actually required knowing the key type should it be an error. Not sure if that would be impractical to implement, but wanted to clarify all the same.

As for the `dict[str, str]()` suggestion, that syntax is not compatible with several Python versions that many of us still need to support for at least another couple years (3.8 doesn't go end-of-life until 2024-10).

Thanks for your consideration of this in any case!

**Comment by erictraut** [comment]
> getting a reference to the `dict_keys` type should not be an error

That would introduce an inconsistency in the way pyright handles unknowns. The inferred type of `{}` is `dict[Unknown, Unknown]`, so `{}.keys()` returns type `dict_keys[Unknown, Unknown]`. You are then passing this value to a `type` call. You've enabled the `reportUnknownArgumentType`, and this argument's type is partially unknown.

If I understand you correctly, you're asking to create an exception in the `reportUnknownArugmentType` rule specifically for the `type` constructor call. Is that correct? In general, we try not to make special-case exceptions like this unless there's a really compelling reason. Such exceptions tend to create confusion, have unpredictable side effects, and make pyright's code difficult to maintain.

**Comment by jab** [comment]
I don't think there should be a special-case exception for the `type` call. Sorry that wasn't clear. On the contrary, I think there should be less special-case logic for the `type` call, e.g.:
```python
x = {}
x.keys().__class__  # Pyright currently does not flag this: good
type(x.keys())      # Pyright currently does flag this: bad
```

The special logic I'm thinking about is, when dealing with a container type where a contained TypeVar is Unknown, don't immediately flag that as an error, but rather wait until some code it's interacting with actually requires the contained TypeVar to be known in order to be proved type-safe before flagging an error. For example:

```python
dict_keys = type({}.keys())  # should not be an error, Unknown key type / value type not relevant on this line
dict_keys.isdisjoint({}.keys(), {1, 2, 3})  # should be an error, need to know key type to prove this call is type safe
```

I'm not sure if this is the best example, but hopefully it clarifies

**Comment by erictraut** [comment]
The statement `x = {}` does not generate any error.

The statement `y = x.keys()` does not generate any error, but y has a partially-unknown type. Depending on how `y` is used, it may generate downstream errors if strict type checking is enabled.

The statement `y.__class__` generates an error if the `reportUnknownMemberType` rule is enabled. This makes sense because `__class__` has the type `type[dict_keys[Unknown, Unknown]]`.

The statement `type(y)` generates an error if the `reportUnknownArgumentType` rule is enabled. This makes sense because `y` has the type `dict_keys[Unknown, Unknown]`.

So `reportUnknownMemberType` and `reportUnknownArgumentType` are reporting errors in a consistent manner here. Any deviation would require special casing and introduce inconsistencies, and I don't think that's advisable.

> when dealing with a container type...

There's no concept of a "container type" in the Python type system. A type checker has no way to distinguish between a "container type" and any other generic type.

> don't immediately flag that as an error

That's what pyright does already. It doesn't flag the error until a value with an unknown type is used. For example, when it is used as the argument in a `type` call expression. Argument types can affect the type analysis of a call expression because of overloads, so an attempt to use a partially-unknown type as an argument can result in incorrect type evaluation for the call expression. That's why `reportUnknownArgumentType` is an important check if you want your code to be strictly typed.

**Comment by jab** [comment]
Oops, I was accidentally picking up a less strict config than I thought when I ran my previous examples. Sorry about that and thanks for the walk-through.

Maybe a better articulation of the idea here is for Pyright to automatically interpret `dict_keys = type({}.keys())` as `dict_keys = type(dict[Any, Any]().keys())` in contexts where no subsequent code is relying on the key type / value type being known, which is the case for the codebase where I hit this, so that users' code can stay cleaner and doesn't have to go out of its way so much to appease the type checker (even in strict mode). Understand if that's not doable, but hopefully this is at least a little clearer now than it was above.

Thanks again for discussing this.
